### PR TITLE
fix: Don't silently ignore duplicates

### LINKED ISSUE #128: Deal with duplicates in a uniform manner
Doesn't cringe:
```
        ServiceCollection.AddNCronJob(n => n
            .AddJob<SimpleJob>(p => p.WithCronExpression("* * * * *"))
            .AddJob<SimpleJob>(p => p.WithCronExpression("* * * * *")));
```

vs

Fails with `InvalidOperationException`
```
        ServiceCollection.AddNCronJob(n => n
            .AddJob<SimpleJob>(p => p.WithCronExpression("* * * * *").WithName("Name"))
            .AddJob<SimpleJob>(p => p.WithCronExpression("* * * * *").WithName("Name")));
```

The code does a lot to avoid duplicated jobDefinitions.

However, the API is a tad inconsistent here with regards to its behavior.
- In the first case, the jobdefinition is silently "replaced"
- In the case case, it throws

I'd be in favor of always failing.
- Adding twice the same jobdefinition is most certainly an error (maybe a hasty copy/paste) and maybe the API should actively protects itself rather than dealing it behind the scene.

### PR DESCRIPTION
**Pull request by linkdotnet**
Fixes #128

### CODE REVIEW

File: src/NCronJob/Registry/JobRegistry.cs [hunk 1]
```diff
@@ -28,13 +28,13 @@ public void Add(JobDefinition jobDefinition)
     {
         AssertNoDuplicateJobNames(jobDefinition.CustomName);
 
-        var isTypeUpdate = allJobs.Any(j => j.JobFullName == jobDefinition.JobFullName);
-        if (isTypeUpdate)
+        if (!allJobs.Add(jobDefinition))
         {
-            Remove(jobDefinition);
+            throw new InvalidOperationException("""
+                                                A duplicated job registration was detected (the same job type with the same parameter and cron expression and not custom name).
```
Comment (nulltoken) [inline src/NCronJob/Registry/JobRegistry.cs:34]: Couldn't we provide in the exception message some info about the jobdefinition (type/name/cron when they exist)?
That would make it easier for the user to quickly figure out what to search for.

Comment (linkdotnet) [reply src/NCronJob/Registry/JobRegistry.cs:34]: The type at least would be something. I will add this tomorrow. Thanks!

Comment (nulltoken) [reply src/NCronJob/Registry/JobRegistry.cs:34]: And maybe refine the error message in `AssertNoDuplicateJobNames` so that it's worded in the same manner ("A duplicated job registration was detected, ...")?

**Comment by linkdotnet** [comment]
Should now include all the changes. After the merge I will rebase `v4`

File: src/NCronJob/Registry/JobRegistry.cs (hunk 1 above)
Comment (nulltoken) [reply src/NCronJob/Registry/JobRegistry.cs]: @linkdotnet So no ❤️ for better messages? :wink:

Comment (linkdotnet) [reply src/NCronJob/Registry/JobRegistry.cs]: Well there is but I will now use ChatGPT for the better :D

Comment (linkdotnet) [reply src/NCronJob/Registry/JobRegistry.cs]: Soooooo. I hope that is a bit better

File: docs/migration/v4.md [hunk 2]
```diff
@@ -44,4 +44,24 @@ The `AddJob` method of the `IRuntimeRegistry` has been renamed to `Register` to
 ```diff
 - runtimeRegistry.AddJob(r => r.AddJob);
 + runtimeRegistry.Register(r => r.AddJob);
+```
+
```
Comment (nulltoken) [inline docs/migration/v4.md:48]: With regards to the RuntimeRegistry, I wonder if we should keep both `Register` and `TryRegister`.

I feel like the impression of choice is biased.
Using `Register` without an inner knowledge of what's at work is running blind with a pair of scissors in hand.

How would you feel about only keeping `TryRegister`?

Independently of your answer, maybe should we add some hint about its existence in the doco. :wink:

File: src/NCronJob/Registry/JobRegistry.cs (hunk 1 above)
Comment (nulltoken) [reply src/NCronJob/Registry/JobRegistry.cs]: Duh. I was unclear in my previous message.

I was referring to maybe add the duplicate type in the error message and maybe reword what AssertNoDuplicateJobNames throws.

Comment (linkdotnet) [reply src/NCronJob/Registry/JobRegistry.cs]: Ah darn me. Sorry. Will do

File: docs/migration/v4.md (hunk 2 above)
Comment (linkdotnet) [reply docs/migration/v4.md:48]: That is a valid point. Another approach is to allow the user to check if a job already exists.
But with the builder pattern, that feels awkward.

Comment (linkdotnet) [reply docs/migration/v4.md:48]: We could also do a mixture: We do the following:
```csharp
bool Register(...);
```

Which is a mix of those two. Where we return `false` if a duplicate is detected. Kind of like a `HashSet`

Comment (linkdotnet) [reply docs/migration/v4.md:48]: See last commit

Comment (nulltoken) [reply docs/migration/v4.md:48]: >bool Register(...);

I don't know. Somehow, I wonder if that wouldn't be somehow worse than the `void` alternative. :-/

The returned bool would certainly be ignored (as it's done, most of the time when using the HashSet ;-) )

And my understanding is that this wouldn't throw anymore. So the thing that user was intended to do wouldn't happen and he would have no clue why.

Comment (linkdotnet) [reply docs/migration/v4.md:48]: Yeah. Just offering "TryRegister" doesn't feel right as well. At least for me.
Maybe we need a "DoesJobExists" - then it is up to the user. But I don't see that is better API wise.

Comment (nulltoken) [reply docs/migration/v4.md:48]: I may be a bit (too much?) obsessed with observabilty, but I think that if we grant the user with the possibility to alter the Registry at runtime, we should provide him/her with ways to be able to understand why something didn't work as expected...

In the end, I think I liked the `TryRegister` option better. At least there was an exception that could be potentially processed.

Comment (linkdotnet) [reply docs/migration/v4.md:48]: Fair. We had some observability planned (well more for the dashboard initially, but hey :D why not reusing it :D )

Before we release `v4` as stable maybe someone has a strong opinion. Let's roll with `TryRegister`

Comment (linkdotnet) [reply docs/migration/v4.md:48]: The only real bummer here is that you always have to specify the `out` variable. That really bums me more out than it should

Comment (linkdotnet) [reply docs/migration/v4.md:48]: Yeah even short: `out _` but even that is a lot. But you are right, the overloads might do the trick. Will work on that

Comment (linkdotnet) [reply docs/migration/v4.md:48]: Said and done! Thanks for the review. I'm loving it

File: src/NCronJob/Registry/JobRegistry.cs [hunk 3]
```diff
@@ -26,13 +26,14 @@ public void Add(JobDefinition jobDefinition)
...
         AssertNoDuplicateJobNames(jobDefinition.CustomName);
 
-        var isTypeUpdate = allJobs.Any(j => j.JobFullName == jobDefinition.JobFullName);
-        if (isTypeUpdate)
+        if (!allJobs.Add(jobDefinition))
         {
-            Remove(jobDefinition);
+            throw new InvalidOperationException(
+                """
+                Job registration conflict detected. Another job with the same type, parameters, or cron expression already exists.
```
Comment (nulltoken) [inline src/NCronJob/Registry/JobRegistry.cs:33]: If possible, add here some metadata about the JobDefinition (type, cron, parameter)

File: src/NCronJob/Registry/JobRegistry.cs [hunk 4]
```diff
@@ -26,13 +26,14 @@ public void Add(JobDefinition jobDefinition)
     {
         AssertNoDuplicateJobNames(jobDefinition.CustomName);
 
-        var isTypeUpdate = allJobs.Any(j => j.JobFullName == jobDefinition.JobFullName);
-        if (isTypeUpdate)
+        if (!allJobs.Add(jobDefinition))
```
Comment (nulltoken) [inline src/NCronJob/Registry/JobRegistry.cs:29]: Maybe reword the exception message thrown by ` AssertNoDuplicateJobNames` so that it better aligns with the wording below".

File: src/NCronJob/Registry/RuntimeJobRegistry.cs [hunk 5]
```diff
@@ -9,9 +10,16 @@ namespace NCronJob;
 public interface IRuntimeJobRegistry
 {
     /// <summary>
-    /// Gives the ability to add a job.
-    /// </summary>
-    void Register(Action<IRuntimeJobBuilder> jobBuilder);
+    /// Tries to register a job with the given configuration. If an exception occurs, it will be returned.
```
Comment (nulltoken) [inline src/NCronJob/Registry/RuntimeJobRegistry.cs:13]:
```suggestion
    /// Tries to register a job with the given configuration.
```

File: docs/advanced/dynamic-job-control.md [hunk 6]
```diff
@@ -27,7 +27,7 @@ To add a job at runtime, leverage the `IRuntimeJobRegistry` interface:
 ```csharp
 app.MapPost("/add-job", (IRuntimeJobRegistry registry) => 
 {
-    registry.AddJob(n => n.AddJob<SampleJob>(p => p.WithCronExpression("* * * * *").WithName("MyName")));
+    registry.Register(n => n.AddJob<SampleJob>(p => p.WithCronExpression("* * * * *").WithName("MyName")));
```
Comment (nulltoken) [inline docs/advanced/dynamic-job-control.md:30]:
```suggestion
    var hasSucceeded = registry.TryRegister(n => n.AddJob<SampleJob>(p => p.WithCronExpression("* * * * *").WithName("MyName")), out Exception exception);
```

Comment (linkdotnet) [reply docs/advanced/dynamic-job-control.md:30]: Oh boy - thanks

File: src/NCronJob/Registry/JobRegistry.cs (hunk 4 above)
Comment (linkdotnet) [reply src/NCronJob/Registry/JobRegistry.cs:29]: Thanks - done

File: src/NCronJob/Registry/JobRegistry.cs (hunk 3 above)
Comment (linkdotnet) [reply src/NCronJob/Registry/JobRegistry.cs]: Added the job type. CRON expression might not be set anyway.

File: src/NCronJob/Configuration/Builder/IRuntimeJobBuilder.cs [hunk 7]
```diff
@@ -29,8 +29,7 @@ public interface IRuntimeJobBuilder
     /// If not set the default time zone is UTC.
     /// </param>
     /// <param name="jobName">Sets the job name that can be used to identify and manipulate the job later on.</param>
-    /// <returns>Returns the <see cref="IRuntimeJobBuilder"/> for chaining.</returns>
-    IRuntimeJobBuilder AddJob(Delegate jobDelegate,
+    void AddJob(Delegate jobDelegate,
```
Comment (nulltoken) [inline src/NCronJob/Configuration/Builder/IRuntimeJobBuilder.cs:32]: nitpick: Don't you want to mention the non longer chaining capability as a breaking change as well?

Comment (linkdotnet) [reply src/NCronJob/Configuration/Builder/IRuntimeJobBuilder.cs:32]: Good point for the migration guide!

File: src/NCronJob/Configuration/Builder/NCronJobOptionBuilder.cs [hunk 8]
```diff
@@ -107,19 +107,11 @@ public NCronJobOptionBuilder AddExceptionHandler<TExceptionHandler>() where TExc
         return this;
     }
 
-    IRuntimeJobBuilder IRuntimeJobBuilder.AddJob<TJob>(Action<JobOptionBuilder>? options)
-    {
-        AddJob<TJob>(options);
-        return this;
-    }
+    void IRuntimeJobBuilder.AddJob<TJob>(Action<JobOptionBuilder>? options) => AddJob<TJob>(options);
```
Comment (nulltoken) [inline src/NCronJob/Configuration/Builder/NCronJobOptionBuilder.cs:110]: same question here.

And (I haven't checked), isn't there any already existing doco that showcase that non longer chaining capability?

File: src/NCronJob/Configuration/Builder/IRuntimeJobBuilder.cs (hunk 7 above)
Comment (linkdotnet) [reply src/NCronJob/Configuration/Builder/IRuntimeJobBuilder.cs:32]: Done

File: src/NCronJob/Configuration/Builder/NCronJobOptionBuilder.cs (hunk 8 above)
Comment (linkdotnet) [reply src/NCronJob/Configuration/Builder/NCronJobOptionBuilder.cs:110]: In the docs for that specific scenario there wasn't any chaining.

File: src/NCronJob/Registry/RuntimeJobRegistry.cs [hunk 9]
```diff
@@ -9,9 +10,16 @@ namespace NCronJob;
...
+    /// Tries to register a job with the given configuration.
+    /// </summary>/param>
+    /// <returns>Returns <c>true</c> if the registration was successful, otherwise <c>false</c>.</returns>
+    bool TryRegister(Action<IRuntimeJobBuilder> jobBuilder);
+
+    /// <summary>
+    /// Tries to register a job with the given configuration.
+    /// </summary>/param>
+    /// <returns>Returns <c>true</c> if the registration was successful, otherwise <c>false</c>.</returns>
+    bool TryRegister(Action<IRuntimeJobBuilder> jobBuilder, [NotNullWhen(false)]out Exception? exception);
```
Comment (nulltoken) [inline src/NCronJob/Registry/RuntimeJobRegistry.cs:22]: Shouldn't this out param be documented through xml doc?

**Review by nulltoken — APPROVED** [review]

**Comment by nulltoken** [comment]
Great job! :clap!

File: src/NCronJob/Registry/RuntimeJobRegistry.cs (hunk 9 above)
Comment (linkdotnet) [reply src/NCronJob/Registry/RuntimeJobRegistry.cs:25]: Indeed - and done.

**Comment by linkdotnet** [comment]
> Great job! 👏

Thanks for sticking around :D that was harder than it looked :D
